File: CSC330/collatz/recursiveJulia/fixedR.jl
struct CollatzValue
    value::BigInt
    c_length::BigInt
end

function Base.show(io::IO, cv::CollatzValue)
    print(io, cv.value, "   ", cv.c_length)
end

function get_length(i::BigInt)
    if i == 1
	return 1
    elseif i % 2 == 0
	return 1 + get_length(div(i, 2))
    else
	return 1 + get_length(i * 3 + 1)
    end
end

function computeCollatzInRange(start::Int, stop::Int)
    top_10 = Vector{CollatzValue}()

    lt(a::CollatzValue, b::CollatzValue) = (a.c_length < b.c_length) || (a.c_length == b.c_length && a.value > b.value)

    start_int = Int(start)
    stop_int = Int(stop)

    for i in start_int:stop_int
        collatz_length = get_length(BigInt(i))
        val = BigInt(i)

        existing_index = findfirst(x -> x.c_length == collatz_length, top_10)
        if existing_index !== nothing
            existing = top_10[existing_index]
            if val < existing.value
                deleteat!(top_10, existing_index)
                push!(top_10, CollatzValue(val, collatz_length))
                sort!(top_10, lt=lt)
            end
        else
            if length(top_10) < 10
                push!(top_10, CollatzValue(val, collatz_length))
                sort!(top_10, lt=lt)
            else
                smallest = top_10[1]
                new_item = CollatzValue(val, collatz_length)
                if lt(smallest, new_item)
                    popfirst!(top_10)
                    push!(top_10, new_item)
                    sort!(top_10, lt=lt)
                end
            end
        end
    end

    sorted_by_length = sort(top_10, by = x -> (-x.c_length, x.value))
    sorted_by_value = sort(top_10, by = x -> x.value)

    return sorted_by_length, sorted_by_value
end

if length(ARGS) < 2
    println("Usage: julia fixed.jl <start> <end>")
    exit(1)
end

start = parse(Int, ARGS[1])
stop = parse(Int, ARGS[2])

sorted_by_length, sorted_by_value = computeCollatzInRange(start, stop)

println("Sorted by length:")
for cv in sorted_by_length
    println(cv)
end

println("Sorted by value:")
for cv in sorted_by_value
    println(cv)
end

File: CSC330/collatz/recursiveC#/fixedR.cs
using System;
using System.Collections.Generic;
using System.Linq;

public struct CollatzValue
{
    public int Value;
    public int CLength;

    public CollatzValue(int value, int cLength)
    {
        Value = value;
        CLength = cLength;
    }

    public override string ToString()
    {
        return $"{Value}   {CLength}";
    }
}

public class CollatzCalculator
{
    public static int GetLength(int i)
    {
	if( i == 1 ){
	    return 1;
	}
	else if( i % 2 == 0) {
	    return 1 + GetLength(i / 2);
	} 
	else {
	    return 1 + GetLength(i * 3 + 1);
	}
    }

    public static (List<CollatzValue> SortedByLength, List<CollatzValue> SortedByValue) ComputeCollatzInRange(int start, int stop)
    {
        List<CollatzValue> top10 = new List<CollatzValue>();

        Func<CollatzValue, CollatzValue, bool> Lt = (a, b) =>
            (a.CLength < b.CLength) || (a.CLength == b.CLength && a.Value > b.Value);

        for (int i = start; i <= stop; i++)
        {
            int collatzLength = GetLength(i);
            int val = i;

            int existingIndex = top10.FindIndex(x => x.CLength == collatzLength);
            if (existingIndex != -1)
            {
                CollatzValue existing = top10[existingIndex];
                if (val < existing.Value)
                {
                    top10.RemoveAt(existingIndex);
                    top10.Add(new CollatzValue(val, collatzLength));
                    top10.Sort((a, b) => Lt(a, b) ? -1 : 1);
                }
            }
            else
            {
                if (top10.Count < 10)
                {
                    top10.Add(new CollatzValue(val, collatzLength));
                    top10.Sort((a, b) => Lt(a, b) ? -1 : 1);
                }
                else
                {
                    CollatzValue smallest = top10[0];
                    CollatzValue newItem = new CollatzValue(val, collatzLength);
                    if (Lt(smallest, newItem))
                    {
                        top10.RemoveAt(0);
                        top10.Add(newItem);
                        top10.Sort((a, b) => Lt(a, b) ? -1 : 1);
                    }
                }
            }
        }

        var sortedByLength = top10.OrderByDescending(x => x.CLength).ThenBy(x => x.Value).ToList();
        var sortedByValue = top10.OrderBy(x => x.Value).ToList();

        return (sortedByLength, sortedByValue);
    }

    public static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: CollatzCalculator <start> <end>");
            Environment.Exit(1);
        }

        int start = int.Parse(args[0]);
        int stop = int.Parse(args[1]);

        var (sortedByLength, sortedByValue) = ComputeCollatzInRange(start, stop);

        Console.WriteLine("Sorted by length:");
        foreach (var cv in sortedByLength)
        {
            Console.WriteLine(cv);
        }

        Console.WriteLine("Sorted by value:");
        foreach (var cv in sortedByValue)
        {
            Console.WriteLine(cv);
        }
    }
}

